File: papyri/graphstore.py
# import json
import logging
import sqlite3
from pathlib import Path
from typing import Any, NamedTuple

log = logging.getLogger("papyri")

# Expanduser is evaluated at module-import time so tests can monkeypatch HOME
# afterwards without invalidating this path.
GLOBAL_PATH = Path("~/.papyri/ingest/papyri.db").expanduser()

# Applied to every connection, old or new.
_PRAGMAS = [
    "PRAGMA foreign_keys = 1",
    "PRAGMA journal_mode = WAL",
    "PRAGMA synchronous = NORMAL",
    "PRAGMA cache_size = -65536",
    "PRAGMA mmap_size = 268435456",
]


class Key(NamedTuple):
    module: str
    version: str
    kind: str
    path: str


class GraphStore:
    """
    Read-only interface to the ingested graphstore.

    The graphstore is a derived cache: a filesystem blob store plus a SQLite
    graph index that the TypeScript ingest pipeline owns exclusively. The
    Python side reads metadata (which documents exist, their forward/back
    references, content digests) but never writes. Schema creation and all
    writes are the responsibility of the TypeScript ingest engine
    (ingest/src/ingest.ts).

    Each document is stored as a CBOR blob on disk, keyed by a 4-tuple
    (package, version, category, identifier). The SQLite database tracks
    which documents exist and which forward/back references connect them.
    """

    def __init__(self, root: Path, link_finder: Any = None) -> None:
        from .config import ensure_dirs

        ensure_dirs()
        p = GLOBAL_PATH
        log.debug("connecting to database %s", p)
        self.conn = sqlite3.connect(str(p))
        self.conn.row_factory = sqlite3.Row
        for pragma in _PRAGMAS:
            self.conn.execute(pragma)

        assert isinstance(root, Path)
        self._root = root
        self._link_finder = link_finder

    def _key_to_path(self, key: Key) -> Path:
        """
        Given a key, return path to the current file.

        Parameters
        ----------
        key : Key

        Returns
        -------
        data_path : Path
        """
        path = self._root
        assert None not in key, key
        for k in key[:-1]:
            path = path / k
        return path / key[-1]

    def _get(self, key: Key) -> bytes:
        assert isinstance(key, Key)
        return self._key_to_path(key).read_bytes()

    def _get_backrefs(self, key: Key) -> set[Key]:
        rows = self.conn.execute(
            """
            SELECT n_src.package, n_src.version, n_src.category, n_src.identifier
            FROM links
            JOIN nodes AS n_src  ON links.source = n_src.id
            JOIN nodes AS n_dest ON links.dest   = n_dest.id
            WHERE n_dest.package=?
              AND n_dest.version=?
              AND n_dest.category=?
              AND n_dest.identifier=?
            """,
            list(key),
        ).fetchall()
        return {Key(r[0], r[1], r[2], r[3]) for r in rows}

    def get_forwardrefs(self, key: Key) -> set[Key]:
        rows = self.conn.execute(
            """
            SELECT n_dest.package, n_dest.version, n_dest.category, n_dest.identifier
            FROM links
            JOIN nodes AS n_src  ON links.source = n_src.id
            JOIN nodes AS n_dest ON links.dest   = n_dest.id
            WHERE n_src.package=?
              AND n_src.version=?
              AND n_src.category=?
              AND n_src.identifier=?
            """,
            list(key),
        ).fetchall()
        return {Key(r[0], r[1], r[2], r[3]) for r in rows}

    def get_all(self, key: Key) -> tuple[bytes, set[Key], set[Key]]:
        a = self._get(key)
        b = self._get_backrefs(key)
        c = self.get_forwardrefs(key)
        return (a, b, c)

    def get_backref(self, key: Key) -> set[Key]:
        return self._get_backrefs(key)

    def get(self, key: Key) -> bytes:
        return self._get(key)

    def get_meta(self, key: Key) -> bytes:
        meta_path = self._root / key.module / key.version / "meta.cbor"
        return meta_path.read_bytes()

    def glob(
        self, pattern: tuple[str | None, str | None, str | None, str | None]
    ) -> list[Key]:
        package, version, category, identifier = pattern
        clauses = []
        params = []
        for col, val in [
            ("package", package),
            ("version", version),
            ("category", category),
            ("identifier", identifier),
        ]:
            if val is not None:
                clauses.append(f"{col}=?")
                params.append(val)
        clauses.append("has_blob=1")
        where = "WHERE " + " AND ".join(clauses)
        rows = self.conn.execute(
            f"SELECT package, version, category, identifier FROM nodes {where}",
            params,
        ).fetchall()
        return [Key(r[0], r[1], r[2], r[3]) for r in rows]

    def get_digest(self, key: Key) -> bytes | None:
        """Return the content digest of the canonical blob bytes for ``key``.

        The digest is a 16-byte BLAKE2b fingerprint recorded at ingest time.
        Returns None if the row exists with no recorded digest, and
        raises ``KeyError`` if the row does not exist at all.
        """
        row = self.conn.execute(
            "SELECT digest FROM nodes"
            " WHERE package=? AND version=? AND category=? AND identifier=?"
            "   AND has_blob=1",
            list(key),
        ).fetchone()
        if row is None:
            raise KeyError(key)
        digest: bytes | None = row["digest"]
        return digest

    def diff_versions(
        self,
        package: str,
        version_a: str,
        version_b: str,
    ) -> list[tuple[str, str, bytes | None, bytes | None]]:
        """Compare two versions of one package by content digest.

        Returns rows ``(category, identifier, digest_a, digest_b)`` for
        every page that exists in at least one of the versions and whose
        digest differs between them. Pages with identical digests are
        omitted.

        - ``digest_a is None``: the page does not exist in ``version_a``
          (i.e. it was added in ``version_b``).
        - ``digest_b is None``: the page does not exist in ``version_b``
          (i.e. it was removed).
        - both not None and different: the page was modified.

        Callers that only care about one ``category`` (module / docs /
        examples / …) can filter the returned list directly.
        """
        # Emulate FULL OUTER JOIN with the union of two LEFT JOINs so we
        # don't depend on SQLite >= 3.39 (Python's bundled sqlite3 version
        # varies by platform/distribution).
        sql = """
            SELECT category, identifier, digest_a, digest_b
            FROM (
                SELECT a.category   AS category,
                       a.identifier AS identifier,
                       a.digest     AS digest_a,
                       b.digest     AS digest_b
                FROM (
                    SELECT category, identifier, digest
                    FROM nodes
                    WHERE package=? AND version=? AND has_blob=1
                ) a
                LEFT JOIN (
                    SELECT category, identifier, digest
                    FROM nodes
                    WHERE package=? AND version=? AND has_blob=1
                ) b USING (category, identifier)
                UNION
                SELECT b.category, b.identifier, a.digest, b.digest
                FROM (
                    SELECT category, identifier, digest
                    FROM nodes
                    WHERE package=? AND version=? AND has_blob=1
                ) b
                LEFT JOIN (
                    SELECT category, identifier, digest
                    FROM nodes
                    WHERE package=? AND version=? AND has_blob=1
                ) a USING (category, identifier)
                WHERE a.identifier IS NULL
            )
            WHERE digest_a IS NOT digest_b
            ORDER BY category, identifier
        """
        params = [
            package,
            version_a,
            package,
            version_b,
            package,
            version_b,
            package,
            version_a,
        ]
        rows = self.conn.execute(sql, params).fetchall()
        return [
            (r["category"], r["identifier"], r["digest_a"], r["digest_b"]) for r in rows
        ]

File: ingest/src/graphstore.ts
/**
 * Write-capable GraphStore: SQLite graph index + filesystem blob store.
 *
 * Mirrors papyri/graphstore.py — same schema, same on-disk layout, same
 * digest algorithm so digests computed here are byte-identical to those
 * produced by the Python tool.
 *
 * Layout: <root>/<module>/<version>/<kind>/<identifier>
 * SQLite: <root>/papyri.db  (nodes + links tables)
 */

import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { blake2b } from "@noble/hashes/blake2b.js";
import DatabaseType from "better-sqlite3";
import Database from "better-sqlite3";

const DIGEST_SIZE = 16;

// ---------------------------------------------------------------------------
// Key
// ---------------------------------------------------------------------------

export interface Key {
  module: string;
  version: string;
  kind: string;
  path: string;
}

export function keyEq(a: Key, b: Key): boolean {
  return a.module === b.module && a.version === b.version && a.kind === b.kind && a.path === b.path;
}

export function keyStr(k: Key): string {
  return `${k.module}/${k.version}/${k.kind}/${k.path}`;
}

// ---------------------------------------------------------------------------
// Schema
//
// Single source of truth: `ingest/migrations/*.sql`. The Cloudflare D1
// path applies the same files via `wrangler d1 migrations apply`
// (`viewer/wrangler.toml` points `migrations_dir = "../ingest/migrations"`).
//
// Two consumers, two ways the schema can reach the GraphStore constructor:
//
//   1. The `papyri-ingest` CLI runs from `dist/cli.js` → `dist/graphstore.js`.
//      `import.meta.url` points at the actual file on disk, so the lazy
//      disk loader below resolves `../migrations/*.sql` correctly.
//
//   2. The viewer's SSR bundle is produced by Vite, which inlines
//      `graphstore.ts` into `dist/server/chunks/<hash>.mjs`. There the
//      `import.meta.url` is a chunk path with no sibling `migrations/`
//      dir. To handle that, callers can pass `schemaSql` explicitly
//      (e.g. via Vite's `?raw` import) and skip the disk loader.
// ---------------------------------------------------------------------------

function migrationsDir(): string {
  // Resolved lazily because this module is loaded transitively by viewer
  // code that runs inside the Cloudflare prerender miniflare worker,
  // where `import.meta.url` is not always a `file:` URL. The
  // `new GraphStore()` constructor that needs this path never runs in
  // that context.
  const here = dirname(fileURLToPath(import.meta.url));
  return join(here, "..", "migrations");
}

function loadSchemaFromDisk(): string {
  const dir = migrationsDir();
  const files = readdirSync(dir)
    .filter((f) => f.endsWith(".sql"))
    .sort(); // lexicographic so 0000_*.sql lands before 0001_*.sql
  return files.map((f) => readFileSync(join(dir, f), "utf8")).join("\n");
}

function splitStatements(sql: string): string[] {
  // SQLite `--` line comments are stripped before splitting so a comment
  // ending in `;` doesn't confuse the splitter. We don't use string
  // literals containing `--` or `;` in the migrations themselves; if that
  // ever changes, swap this for a real tokenizer.
  const stripped = sql
    .split("\n")
    .filter((line) => !line.trim().startsWith("--"))
    .join("\n");
  return stripped
    .split(";")
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

const PRAGMAS = [
  "PRAGMA foreign_keys = 1",
  "PRAGMA journal_mode = WAL",
  "PRAGMA synchronous = NORMAL",
  "PRAGMA cache_size = -65536",
  "PRAGMA mmap_size = 268435456",
];

// ---------------------------------------------------------------------------
// GraphStore
// ---------------------------------------------------------------------------

export interface GraphStoreOptions {
  /**
   * Schema SQL applied when the SQLite file is first created. Combined
   * `migrations/*.sql` content; statements are split on `;` and exec'd
   * in order. Pass this when the caller's runtime can't reach the
   * on-disk migrations dir — e.g. a Vite-bundled SSR worker. The CLI
   * can omit it; the disk loader handles `dist/cli.js` correctly.
   */
  schemaSql?: string;
}

export class GraphStore {
  private db: DatabaseType.Database;
  private root: string;

  constructor(ingestDir: string, options: GraphStoreOptions = {}) {
    this.root = ingestDir;
    const dbPath = join(ingestDir, "papyri.db");
    const isNew = !existsSync(dbPath);
    mkdirSync(ingestDir, { recursive: true });

    this.db = new Database(dbPath);
    for (const p of PRAGMAS) this.db.exec(p);

    if (isNew) {
      const sql = options.schemaSql ?? loadSchemaFromDisk();
      for (const stmt of splitStatements(sql)) this.db.exec(stmt);
    }
  }

  // -------------------------------------------------------------------------
  // Internal helpers
  // -------------------------------------------------------------------------

  private keyToPath(key: Key): string {
    return join(this.root, key.module, key.version, key.kind, key.path);
  }

  private maybeInsertNode(
    key: Key,
    opts: { hasBlob?: boolean; digest?: Uint8Array | null } = {},
  ): number {
    const { hasBlob = false, digest = null } = opts;
    // INSERT OR IGNORE; then SELECT to get the id either way.
    let row = this.db
      .prepare(
        "INSERT OR IGNORE INTO nodes(package, version, category, identifier)" +
          " VALUES (?, ?, ?, ?) RETURNING id",
      )
      .get(key.module, key.version, key.kind, key.path) as { id: number } | undefined;

    if (row === undefined) {
      row = this.db
        .prepare(
          "SELECT id FROM nodes WHERE package=? AND version=? AND category=? AND identifier=?",
        )
        .get(key.module, key.version, key.kind, key.path) as { id: number };
    }

    const nodeId = row.id;
    if (hasBlob) {
      this.db
        .prepare("UPDATE nodes SET has_blob=1, digest=? WHERE id=?")
        .run(digest ?? null, nodeId);
    }
    return nodeId;
  }

  // -------------------------------------------------------------------------
  // Public API
  // -------------------------------------------------------------------------

  /**
   * Store blob bytes under `key`, recording forward-ref links to `refs`.
   * Computes a 16-byte BLAKE2b digest of `bytes` (matching Python's
   * `hashlib.blake2b(bytes, digest_size=16)`).
   */
  put(key: Key, bytes: Uint8Array, refs: Key[]): void {
    const blobPath = this.keyToPath(key);
    mkdirSync(dirname(blobPath), { recursive: true });

    // For non-asset keys, read old forward refs so we can diff them.
    let oldRefs: Set<string>;
    if (key.kind !== "assets" && existsSync(blobPath)) {
      oldRefs = new Set(this.getForwardRefs(key).map(keyStr));
    } else {
      oldRefs = new Set();
    }

    writeFileSync(blobPath, bytes);
    const digest = blake2b(bytes, { dkLen: DIGEST_SIZE });

    const newRefSet = new Set(refs.map(keyStr));
    const addedRefs = refs.filter((r) => !oldRefs.has(keyStr(r)));
    const removedRefStrs = [...oldRefs].filter((s) => !newRefSet.has(s));

    const tx = this.db.transaction(() => {
      const sourceId = this.maybeInsertNode(key, { hasBlob: true, digest });

      const addParams = addedRefs.map(
        (r) => [sourceId, this.maybeInsertNode(r)] as [number, number],
      );
      const addStmt = this.db.prepare("INSERT OR IGNORE INTO links(source, dest) VALUES (?,?)");
      for (const p of addParams) addStmt.run(...p);

      if (removedRefStrs.length > 0) {
        // Look up ids for the removed refs.
        const placeholders = removedRefStrs.map(() => "(?,?,?,?)").join(",");
        const params: string[] = [];
        for (const s of removedRefStrs) {
          const [mod, ver, kind, path] = s.split("/");
          params.push(mod ?? "", ver ?? "", kind ?? "", path ?? "");
        }
        const rows = this.db
          .prepare(
            `SELECT id FROM nodes WHERE (package, version, category, identifier) IN (VALUES ${placeholders})`,
          )
          .all(...params) as { id: number }[];
        const delStmt = this.db.prepare("DELETE FROM links WHERE source=? AND dest=?");
        for (const r of rows) delStmt.run(sourceId, r.id);
      }
    });
    tx();
  }

  /** Write a per-bundle meta.cbor (not tracked in the nodes table). */
  putMeta(module: string, version: string, data: Uint8Array): void {
    const metaPath = join(this.root, module, version, "meta.cbor");
    mkdirSync(dirname(metaPath), { recursive: true });
    writeFileSync(metaPath, data);
  }

  /** Read blob bytes for a key. */
  get(key: Key): Buffer {
    return readFileSync(this.keyToPath(key));
  }

  /** Return all keys with blobs that match the (possibly-null) pattern fields. */
  glob(pattern: Partial<Key>): Key[] {
    const clauses: string[] = [];
    const params: string[] = [];
    for (const [col, field] of [
      ["package", "module"],
      ["version", "version"],
      ["category", "kind"],
      ["identifier", "path"],
    ] as [string, keyof Key][]) {
      const val = pattern[field];
      if (val !== undefined && val !== null) {
        clauses.push(`${col}=?`);
        params.push(val);
      }
    }
    clauses.push("has_blob=1");
    const where = "WHERE " + clauses.join(" AND ");
    const rows = this.db
      .prepare(`SELECT package, version, category, identifier FROM nodes ${where}`)
      .all(...params) as {
      package: string;
      version: string;
      category: string;
      identifier: string;
    }[];
    return rows.map((r) => ({
      module: r.package,
      version: r.version,
      kind: r.category,
      path: r.identifier,
    }));
  }

  /** Return the set of keys that `key` references (forward edges). */
  getForwardRefs(key: Key): Key[] {
    const rows = this.db
      .prepare(
        "SELECT n_dest.package, n_dest.version, n_dest.category, n_dest.identifier " +
          "FROM links " +
          "JOIN nodes AS n_src  ON links.source = n_src.id " +
          "JOIN nodes AS n_dest ON links.dest   = n_dest.id " +
          "WHERE n_src.package=? AND n_src.version=? AND n_src.category=? AND n_src.identifier=?",
      )
      .all(key.module, key.version, key.kind, key.path) as {
      package: string;
      version: string;
      category: string;
      identifier: string;
    }[];
    return rows.map((r) => ({
      module: r.package,
      version: r.version,
      kind: r.category,
      path: r.identifier,
    }));
  }

  /** Return the set of keys that link TO `key` (back edges). */
  getBackRefs(key: Key): Key[] {
    const rows = this.db
      .prepare(
        "SELECT n_src.package, n_src.version, n_src.category, n_src.identifier " +
          "FROM links " +
          "JOIN nodes AS n_src  ON links.source = n_src.id " +
          "JOIN nodes AS n_dest ON links.dest   = n_dest.id " +
          "WHERE n_dest.package=? AND n_dest.version=? AND n_dest.category=? AND n_dest.identifier=?",
      )
      .all(key.module, key.version, key.kind, key.path) as {
      package: string;
      version: string;
      category: string;
      identifier: string;
    }[];
    return rows.map((r) => ({
      module: r.package,
      version: r.version,
      kind: r.category,
      path: r.identifier,
    }));
  }

  close(): void {
    this.db.close();
  }
}
